crates/cranelift/src/wasmfx/optimized.rs

```diff
@@ -1463,17 +1506,158 @@ pub(crate) fn typed_continuations_store_payloads<'a>(
     }
 }
 
-/// We only use this when we want to get the current continuation for the
-/// purpose of suspending. Thus, if there is no such continuation (because we
-/// are on the main stack), we trap with an unhandled tag error.
-pub(crate) fn typed_continuations_load_continuation_reference<'a>(
+/// This function generates code that searches for a handler for `tag_address`,
+/// which must be a `*mut VMTagDefinition`. The search walks up the chain of
+/// continuations beginning at `start`.
+///
+/// The returned values are:
+/// 1. The continuation whose *parent* (another continuation or the
+///    main stack) handles `tag_address`.
+/// 2. The index of the handler in the parent's HandlerList.
+///
+/// In pseudo-code, the generated code's behavior can be expressed as
+/// follows:
+///
+/// chain_link = start
+/// while !chain_link.is_main_stack() {
+///   contref = chain_link.get_contref()
+///   parent_link = contref.parent
+///   parent_csi = parent_link.get_common_stack_information();
+///   handlers = parent_csi.handlers;
```

Comment:
This doesn't seem right to me. The handlers should sit on the current continuation reference. If the current one doesn't handle the suspension, then we try the parent, and so forth.

Reply:
I'm not sure if I understand what you are saying.

Consider the following example:
```
  ...
  
   ^
   | parent 
   
  c1
  
   ^
   | parent
   
  c2
```

where `$c1` executed `(resume $ct (on $t switch) ... (local.get $c2))` and `$c2` is currently executing.

In the current design in this PR, continuation `$c1` would have a handler list _l_ with a `switch` entry for `$t` (as well as all other tags potentially occurring in the `resume` above).
Continuation `$c2` is currently running and has no handler list.

Are you suggesting that the handler list should be attached to `$c2` instead, meaning that each continuation stores what is handled by its immediate parent?

The disadvantage of that approach over the current one is that it requires additional work on `switch` instructions:
In the example above, if `$c2` then executes `(switch $ct' $t (local.get $c3))` for some `$c3`, we would have to update the handler lists such that _l_ becomes the handler list of `$c3` (and the handler list of `$c2` is cleared for housekeeping).

This work isn't required if the handler list _l_ is associated with `$c1` instead (and stays there on switch).

Reply:
The semantics of `switch` is similar to the semantics of a deep handler, in the sense that the handler stays in place after resume. So any proper design should not have to modify the handler on `switch`.